In [1]:
library(ggplot2)
library(reshape2)
library(RColorBrewer)
suppressMessages(library(dplyr))
library(stringr)
suppressMessages(library(tidyr))
theme_set(theme_bw())
library(scales)
options(repr.plot.width=7, repr.plot.height=4)
isotypes = c('Ala', 'Arg', 'Asn', 'Asp', 'Cys', 'Gln', 'Glu', 'Gly', 'His', 'Ile', 'iMet', 'Leu', 'Lys', 'Met', 'Phe', 'Pro', 'Ser', 'Thr', 'Trp', 'Tyr', 'Val')

paired_positions = c('X1.72'='1:72', 'X2.71'='2:71', 'X3.70'='3:70', 'X4.69'='4:69', 'X5.68'='5:68', 'X6.67'='6:67', 'X7.66'='7:66', 'X8.14'='*8:14', 'X9.23'='*9:23', 'X10.25'='10:25', 'X10.45'='*10:45', 'X11.24'='11:24', 'X12.23'='12:23', 'X13.22'='13:22', 'X15.48'='*15:48','X18.55'='*18:55', 'X19.56'='*19:56', 'X22.46'='*22:46', 'X26.44'='*26:44', 'X27.43'='27:43', 'X28.42'='28:42', 'X29.41'='29:41', 'X30.40'='30:40', 'X31.39'='31:39', 'X49.65'='49:65', 'X50.64'='50:64', 'X51.63'='51:63', 'X52.62'='52:62', 'X53.61'='53:61', 'X54.58'='*54:58')
paired_identities = c('GC', 'AU', 'UA', 'CG', 'GU', 'UG', 'PurinePyrimidine', 'PyrimidinePurine', 'StrongPair', 'WeakPair', 'AminoKeto', 'KetoAmino', 'Wobble', 'Paired', 'Bulge', 'Mismatched', 'Absent')

single_positions = c('X8'='8', 'X9'='9', 'X14'='14', 'X15'='15', 'X16'='16', 'X17'='17', 'X17a'='17a', 'X17b'='17b', 'X18'='18', 'X19'='19', 'X20'='20', 'X20a'='20a', 'X20b'='20b', 'X21'='21', 'X26'='26', 'X32'='32', 'X33'='33', 'X34'='34', 'X35'='35', 'X36'='36', 'X37'='37', 'X38'='38', 'X44'='44', 'X45'='45', 'X46'='46', 'X47'='47', 'X48'='48', 'X54'='54', 'X55'='55', 'X56'='56', 'X57'='57', 'X58'='58', 'X59'='59', 'X60'='60', 'X73'='73')
single_identities = c('A', 'C', 'G', 'U', 'Purine', 'Pyrimidine', 'Weak', 'Strong', 'Amino', 'Keto', 'B', 'D', 'H', 'V', 'Absent')

colors = c('A'='gray20', 'C'='gray20', 'G'='gray20', 'U'='gray20', 'Absent'='gray20', 'Purine'='gray20', 'Pyrimidine'='gray20', 'Weak'='gray20', 'Strong'='gray20', 'Amino'='gray20', 'Keto'='gray20', 'B'='gray50', 'D'='gray50', 'H'='gray50', 'V'='gray50', 'GC'='gray20', 'AU'='gray20', 'UA'='gray20', 'CG'='gray20', 'GU'='gray20', 'UG'='gray20', 'PurinePyrimidine'='gray20', 'PyrimidinePurine'='gray20', 'StrongPair'='gray20', 'WeakPair'='gray20', 'AminoKeto'='gray20', 'KetoAmino'='gray20', 'Wobble'='gray20', 'Paired'='gray20', 'Bulge'='gray20', 'Mismatched'='gray20')
fills = c('A'='#ffd92f', 'C'='#4daf4a', 'G'='#e41a1c', 'U'='#377eb8', 'Purine'='#ff8300', 'Pyrimidine'='#66c2a5', 'Weak'='#b3de69', 'Strong'='#fb72b2', 'Amino'='#c1764a', 'Keto'='#b26cbd', 'B'='#e5c494', 'D'='#ccebd5', 'H'='#ffa79d', 'V'='#a6cdea', 'Absent'='gray60', 'Mismatched'='gray30', 'Paired'='#ffffcc')

In [2]:
load('best-freqs.RData')
load('clade-isotype-specific.RData')
load('isotype-specific.RData')
load('consensus-IDEs.RData')
load('clade-isotype-specific-freqs.RData')

In [3]:
identities = read.delim('identities.tsv', sep='\t')
identities$quality = as.logical(identities$quality)
identities$restrict = as.logical(identities$restrict)
identities = identities %>% mutate(quality=quality & !restrict)

# Introduction

Strategy 1: find rejected (a) consensus, (b) isotype-specific, or (c) clade-isotype-specific elements. Set a cutoff at 95% presence, but fails (i) isotype check, (ii) clade check, or (iii) species check. Spin in previous work, if any. Bonus points if previous work _conflicts_ with consensus definition.

Strategy 2: highlight instances where previous work conflicts.

Some of these will have been covered in `first-pass-consensus`.

# Strategy 1A

## Find targets

In [4]:
code_groups = c('A'=1, 'C'=1, 'G'=1, 'U'=1, 'Absent'=1, 
                'Purine'=2, 'Pyrimidine'=2,
                'Weak'=3, 'Strong'=3, 'Amino'=3, 'Keto'=3,
                'B'=4, 'D'=4, 'H'=4, 'V'=4,
                'GC'=1, 'AU'=1, 'UA'=1, 'CG'=1, 'GU'=1, 'UG'=1,
                'StrongPair'=2, 'WeakPair'=2, 'Wobble'=2,
                'PurinePyrimidine'=3, 'PyrimidinePurine'=3, 'AminoKeto'=3, 'KetoAmino'=3,
                'Paired'=4, 'Mismatched'=4, 'Bulge'=4)

cutoff_freqs = data.frame()
for (cutoff in c(0.5, 0.6, 0.7, 0.8, 0.9, 0.95, 0.99, 1.0)) {
  df = clade_iso_ac_freqs %>%
    group_by(positions, variable) %>%
    summarize(count=sum(value), freq=sum(value)/sum(total)) %>%
    filter(freq >= cutoff) %>%
    mutate(cutoff=as.character(cutoff)) %>%
    select(positions, variable, freq, cutoff) %>%
    group_by(positions) %>%
    arrange(code_groups[variable], desc(freq)) %>%
    filter(row_number(positions) == 1)
  if (nrow(cutoff_freqs) == 0) cutoff_freqs = df
  else cutoff_freqs = rbind(cutoff_freqs, df)
}

In [5]:
cutoff_freqs %>% 
  filter(positions %in% c(names(single_positions), names(paired_positions))) %>%
  select(positions, cutoff, variable) %>%
  spread(cutoff, variable)

Source: local data frame [64 x 8]
Groups: positions [64]

   positions            `0.5`            `0.6`            `0.7`
*      <chr>           <fctr>           <fctr>           <fctr>
1     X10.25               GC               GC               GC
2     X10.45       Mismatched       Mismatched       Mismatched
3     X11.24               CG               CG PyrimidinePurine
4     X12.23 PyrimidinePurine PyrimidinePurine           Paired
5     X13.22 PyrimidinePurine           Paired               NA
6        X14                A                A                A
7        X15                G                G                G
8     X15.48               GC               GC               GC
9        X16                U                U                U
10       X17           Absent           Absent           Absent
# ... with 54 more rows, and 4 more variables: `0.8` <fctr>, `0.9` <fctr>,
#   `0.95` <fctr>, `0.99` <fctr>

In [6]:
consensus

# A tibble: 26 × 2
   positions identity
       <chr>    <chr>
1     X10.25   Paired
2     X11.24   Paired
3        X14        A
4        X15        V
5        X18        V
6        X19   Purine
7        X21        H
8      X2.71   Paired
9     X28.42   Paired
10    X30.40   Paired
# ... with 16 more rows

For strategy 1A, some good candidates here: 
- R9 (95%)
- Y11:R24 (99%)
- R15:Y48 (95%)
- Y16 (95%)
- G18, U55, and G18:U55 (99%)
- G19:C56 (99%)
- A21 and A14/A21 (99%)
- U33 (99%)
- R37 (99%)
- R46 (99%)
- G53:C61 (99%)
- U54:A58 (99%)
- U55 (99%)
- R57 (99%)
- Y60 (95%)

## R9 and R46

### What's known

- Marck and Grosjean have a purine for iMet and no consensus for elongator, though they are mostly R or V. R9 is also conserved in initiators for archaea/bacteria.
- 9-12-23 is a tertiary interaction. [Gautheret et al. (1995)](http://dx.doi.org/10.1006/jmbi.1995.0200) has these frequencies from the Sprinzl 1991 database: ![9-12-23 frequency matrix](figures/9-12-23-gautheret.png)

- Position 9 is known to be modified with m$^1$G (along with 37) in a wide range of eukaryotes. 

### Our data

- Here's our frequencies:

In [7]:
table(paste0(identities[identities$quality, ]$X12, ':', identities[identities$quality, ]$X23), identities[identities$quality, ]$X9)

     
         -    A    C    G    U
  A:A    0    5    2    6    0
  A:C    0   21    0  199    0
  A:G    0    2    1   10    0
  A:U    0  434  148   57    4
  C:A    0    6    0   55    1
  C:C    0    1    0    2    0
  C:G    0  200   46 5731   26
  C:U    0    0    0    3    0
  G:A    0    1    0    2    0
  G:C    0 1363    7 3371    5
  G:G    0    0    0    8    0
  G:U    0   12    1   29    0
  U:A    1 5501    7 1419   29
  U:C    0    1    0    2    0
  U:G    0   20    0   50    0
  U:U    0    4    0    1    0

#### Which isotypes/clades/species fail the consensus checks?

In [20]:
## Clade/isotype check
best_freqs %>% filter(positions == 'X9') %>% group_by(clade, isotype) %>%
  summarize(status=sum((variable %in% c("G", "A", "Purine"))*count)/sum(count) == 1) %>%
  filter(!status)

## Species check
identities %>% select(clade, species, isotype, X9) %>% group_by(clade, species, isotype) %>%
  summarize(status=sum(X9 %in% c("G", "A", "Purine")/n()) >= 0.1,
            freq=sum(X9 %in% c("G", "A", "Purine")), ntRNAs=n()) %>%
  filter(!status)

Source: local data frame [6 x 3]
Groups: clade [5]

          clade isotype status
          <chr>   <chr>  <lgl>
1       Insecta     Glu  FALSE
2       Insecta     His  FALSE
3      Mammalia     His  FALSE
4 Spermatophyta     Glu  FALSE
5  Streptophyta     Glu  FALSE
6    Vertebrata     His  FALSE

Source: local data frame [56 x 6]
Groups: clade, species [56]

      clade                        species isotype status  freq ntRNAs
     <fctr>                         <fctr>  <fctr>  <lgl> <int>  <int>
1     Fungi                botrCine_B05_10     Tyr  FALSE     0      5
2     Fungi                 crypGatt_WM276     Gly  FALSE     0      2
3     Fungi        crypNeof_VAR_GRUBII_H99     Gly  FALSE     0      2
4     Fungi crypNeof_VAR_NEOFORMANS_B_3501     Gly  FALSE     0      2
5     Fungi  crypNeof_VAR_NEOFORMANS_JEC21     Gly  FALSE     0      2
6   Insecta                            dm6     His  FALSE     0      5
7   Insecta                            dp4     His  FALSE     0      5
8   Insecta                        droYak2     His  FALSE     0      5
9  Mammalia                        ailMel1     His  FALSE     0      8
10 Mammalia                        balAcu1     His  FALSE     0      8
# ... with 46 more rows

### What's new

The 9:12:23 ratios are a bit different. But it shows the same thing: there is some selectivity for what base triples are allowed, but enough tolerance that there are a few interactions that persist.

The default hypothesis is that the 9-23 interaction has a conserved interaction type that explains the frequencies shown. We can marginalize as such because 12:23 is always a WC pair. This is not the case - although some (R:R) have a a strong trans WC-Hoogsteen pair of hydrogen bonds, others (C:R) do not. The hoogsteen side of C is just two stable carbons unlikely to hydrogen bond.

Examining the frequencies _without_ histidine shows that the C9:U23 is conserved in His, and A9:U23 is conserved in Asp, for some clades. (See tertiary interactions figure).

In [9]:
table(identities[identities$quality, ]$X9, identities[identities$quality, ]$X23)

   
       -    A    C    G    U
  -    0    1    0    0    0
  A    0 5513 1386  222  450
  C    0    9    7   47  149
  G    0 1482 3574 5799   90
  U    0   30    5   26    4

So here's what we know.
- 12:23 is almost always WC paired, so we can marginalize it.
- 9 is typically a purine, though there are isotype- and clade-specific exceptions. For example, His and Asp deviate in 4 clades.
    - In the His case, C and U can interact via the 2-carbonyl [amino pairing](http://www.columbuslabs.org/wp-content/uploads/2008/03/basepairs.pdf).
- 9:23 is thought to be a _trans_ interaction with two hydrogen bonds.

**Conclusion 1**: Position 9 was thought to be a purine. Instead, it varies by isotype and clade. Much of the variation from R9 can be explained by looking at 9:23 as an clade/isotype discriminating tertiary interaction.

We can leave it at that. But we should also look into compensatory interactions if 9:23 is disrupted.

## Covariation for core D stem 3d pairs

#### tRNA covariation frequencies for 9:23, 22:46, and 10:45

In [10]:
df = table(paste0(identities[identities$quality, ]$X9, ':', identities[identities$quality, ]$X23),
           paste0(identities[identities$quality, ]$X46, ':', identities[identities$quality, ]$X22, ' / ', identities[identities$quality, ]$X45, ':', identities[identities$quality, ]$X10),
           identities[identities$quality, ]$isotype)

In [11]:
as.data.frame(df) %>% group_by(Var3) %>% filter(Freq > 50)

Source: local data frame [48 x 4]
Groups: Var3 [21]

     Var1      Var2   Var3  Freq
   <fctr>    <fctr> <fctr> <int>
1     A:A G:G / G:G    Ala  1178
2     G:C G:G / G:G    Ala    84
3     G:C A:A / G:G    Arg   252
4     G:G A:A / G:G    Arg   821
5     G:C G:G / G:G    Arg   364
6     G:G A:A / G:G    Asn    59
7     A:A G:G / G:G    Asn   106
8     G:C G:G / G:G    Asn   388
9     A:U A:G / G:G    Asp   301
10    A:A G:G / G:G    Cys   777
# ... with 38 more rows

This is messy. For each isotype there may be a different tertiary interaction compensating for the lack of A:A. It seems that purine:purine interactions are unusually enriched among all three of these except in valine. 

![classical interaction structure](figures/3d-interactions-oliva.png)

#### Average number of purine:purine interactions by isotype and clade

In [12]:
RRs = c("A:A", "A:G", "G:G", "G:A")
identities %>% select(isotype, clade, quality, X9.23, X22.46, X10.45, X26.44) %>% 
  filter(quality) %>%
  rowwise() %>% 
  mutate(nRR=(X9.23 %in% RRs) + (X22.46 %in% RRs) + (X10.45 %in% RRs) + (X26.44 %in% RRs)) %>%
  group_by(isotype, clade) %>% 
  summarize(nRR=signif(mean(nRR), 3)) %>%
  spread(isotype, nRR)

Warning message:
: Grouping rowwise data frame strips rowwise nature

# A tibble: 7 × 22
          clade   Ala   Arg   Asn   Asp   Cys   Gln   Glu   Gly   His   Ile
*        <fctr> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
1         Fungi  2.88  3.23  3.57  3.02  3.09  1.46  1.89  1.93  2.87  3.37
2       Insecta  4.00  3.61  3.00  2.00  4.00  2.00  3.25  2.43  2.20  3.82
3      Mammalia  3.97  3.54  2.96  2.12  3.97  1.98  2.96  2.62  2.05  3.58
4      Nematoda  3.98  3.58  2.94  2.00  4.00  1.93  2.75  2.49  1.94  3.52
5 Spermatophyta  4.00  3.48  3.79  2.04  3.85  2.67  3.55  3.04  3.00  4.00
6  Streptophyta  3.98  3.47  3.73  1.98  3.59  2.40  3.41  3.26  3.00  3.99
7    Vertebrata  3.95  3.48  2.92  2.01  3.92  1.98  2.96  2.35  1.97  3.62
# ... with 11 more variables: iMet <dbl>, Leu <dbl>, Lys <dbl>, Met <dbl>,
#   Phe <dbl>, Pro <dbl>, Ser <dbl>, Thr <dbl>, Trp <dbl>, Tyr <dbl>, Val <dbl>

#### Average number of purine:purine interactions by position/isotype

In [13]:
RRs = c("A:A", "A:G", "G:G", "G:A")
identities %>% select(isotype, quality, X9.23, X22.46, X10.45, X26.44) %>% 
  filter(quality) %>%
  rowwise() %>% 
  mutate(X9.23=X9.23 %in% RRs, X22.46=X22.46 %in% RRs, X10.45=X10.45 %in% RRs, X26.44=X26.44 %in% RRs) %>%
  gather(position, RR, X9.23, X22.46, X10.45, X26.44, -isotype, -quality) %>%
  group_by(isotype, position) %>%
  summarize(RR=round(mean(RR), digits=1)) %>%
  spread(isotype, RR)

# A tibble: 4 × 22
  position   Ala   Arg   Asn   Asp   Cys   Gln   Glu   Gly   His   Ile  iMet
*    <chr> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
1   X10.45   1.0   1.0   1.0   1.0   1.0   1.0   1.0   0.9   1.0   1.0   0.8
2   X22.46   0.9   1.0   1.0   1.0   1.0   0.8   0.9   0.9   0.9   1.0   1.0
3   X26.44   1.0   0.9   0.9   0.1   0.9   0.0   0.2   0.1   0.2   0.9   1.0
4    X9.23   0.9   0.6   0.3   0.2   1.0   0.1   0.8   0.6   0.2   0.7   0.0
# ... with 10 more variables: Leu <dbl>, Lys <dbl>, Met <dbl>, Phe <dbl>,
#   Pro <dbl>, Ser <dbl>, Thr <dbl>, Trp <dbl>, Tyr <dbl>, Val <dbl>

#### Average number of interactions containing a purine by position/isotype

In [14]:
RRs = c("A:A", "A:G", "G:G", "G:A", "G:C", "C:G", "A:C", "C:A", "U:G", "G:U", "A:U", "U:A")
identities %>% select(isotype, quality, X9.23, X22.46, X10.45, X26.44) %>% 
  filter(quality) %>%
  rowwise() %>% 
  mutate(X9.23=X9.23 %in% RRs, X22.46=X22.46 %in% RRs, X10.45=X10.45 %in% RRs, X26.44=X26.44 %in% RRs) %>%
  gather(position, RR, X9.23, X22.46, X10.45, X26.44, -isotype, -quality) %>%
  group_by(isotype, position) %>%
  summarize(RR=round(mean(RR), digits=1)) %>%
  spread(isotype, RR)

# A tibble: 4 × 22
  position   Ala   Arg   Asn   Asp   Cys   Gln   Glu   Gly   His   Ile  iMet
*    <chr> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
1   X10.45     1     1     1   1.0     1   1.0   1.0   0.9   1.0     1     1
2   X22.46     1     1     1   1.0     1   1.0   1.0   1.0   1.0     1     1
3   X26.44     1     1     1   0.8     1   0.1   0.4   1.0   0.4     1     1
4    X9.23     1     1     1   1.0     1   1.0   1.0   1.0   0.5     1     1
# ... with 10 more variables: Leu <dbl>, Lys <dbl>, Met <dbl>, Phe <dbl>,
#   Pro <dbl>, Ser <dbl>, Thr <dbl>, Trp <dbl>, Tyr <dbl>, Val <dbl>

#### Are non-purine-containing interactions enriched for some other type of interaction?

In [15]:
identities %>% select(isotype, quality, X9.23, X22.46, X10.45, X26.44) %>% 
  filter(quality) %>%
  gather(position, identity, X9.23, X22.46, X10.45, X26.44, -isotype, -quality) %>%
  filter(!(identity %in% c(RRs, "G:-", "A:-"))) %>%
  group_by(position, isotype, identity) %>%
  summarize(count=n()) %>%
  filter(count > 20) %>%
  spread(identity, count, fill=0)

Warning message:
: attributes are not identical across measure variables; they will be dropped

Source: local data frame [9 x 7]
Groups: position, isotype [9]

  position isotype `C:-` `C:U` `U:-` `U:C` `U:U`
*    <chr>  <fctr> <dbl> <dbl> <dbl> <dbl> <dbl>
1   X10.45     Leu    44     0     0     0     0
2   X10.45     Ser    86     0   149     0     0
3   X26.44     Asp     0     0     0     0    47
4   X26.44     Gln     0    28     0   585    99
5   X26.44     Glu     0     0     0   515    94
6   X26.44     Gly     0     0     0     0    25
7   X26.44     His     0     0     0   174     0
8   X26.44     Ile     0     0     0    36     0
9    X9.23     His     0   149     0     0     0

They are not - but they are isotype-specific.

**Conclusion**: Tertiary interactions within this "tRNA core region" along the D stem are enriched in purines and purine-purine interactions, though the extent varies by isotype and position. Exceptions are isotype-specific.

## G18, U55, and G18:U55

#### Number of tRNAs without G18:U55

In [16]:
identities %>% select(clade, species, quality, X18, X55, X18.55) %>%
  filter(quality & X18.55 != "G:U") %>% nrow()

[1] 288

#### Which isotypes/clades/species fail the consensus checks?

In [21]:
## Clade/isotype check
best_freqs %>% filter(positions == 'X18.55') %>% group_by(clade, isotype) %>%
  summarize(status=sum((variable == "GU")*count)/sum(count) == 1) %>%
  filter(!status)

## Species check
identities %>% select(clade, species, isotype, quality, X18.55) %>%
  filter(quality) %>%
  group_by(clade, species, isotype) %>%
  summarize(status=sum((X18.55 == "G:U")/n()) >= 0.1,
            freq=sum(X18.55 == "G:U"), tRNAs=n()) %>%
  filter(!status)

Source: local data frame [2 x 3]
Groups: clade [2]

       clade isotype status
       <chr>   <chr>  <lgl>
1   Nematoda     Asn  FALSE
2 Vertebrata     Gly  FALSE

Source: local data frame [2 x 6]
Groups: clade, species [2]

   clade            species isotype status  freq tRNAs
  <fctr>             <fctr>  <fctr>  <lgl> <int> <int>
1  Fungi flamVelu_KACC42780     Pro  FALSE     0     1
2  Fungi      valsMali_03_8     Pro  FALSE     0     1

The isotype/clade check fails for nematode Asn at 88% and vertebrate Gly at 84%. The two fungi listed here are rare exceptions to the rule. 

Earlier, we had a vertebrate check fail because of the methods used to build the quality set. I've relaxed the species check constraints to 10% to better identify IDEs even when there are massively amplified tRNAs. Curiously, in danRer10, gasAcu1, and oryLat2, the majority of the non-G:U tRNAs contain a G:C. It's pretty likely that, given the high copy number, many of these tRNAs are actually pseudogenes. In addition, there are a decent amount of G18:U55 tRNAs to actually drive translation.